### ISSUE TITLE
avoid guarding on max() unnecessarily

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#149635]
here's a repro. theoretically the code below should not require a recompile. We are conditionally padding, producing an output tensor of shape max(input_size, 16). Instead though, we specialize on the pad value, and produce separate graphs for the `size_16` and `size_greater_than_16` cases
```
import torch

@torch.compile(backend="eager")
def f(x):
    padded_size = max(x.shape[0], 16)
    padded_tensor = torch.ones(padded_size, *x.shape[1:])
    return padded_tensor + x.sum()


x = torch.arange(15)
torch._dynamo.mark_dynamic(x, 0)
out = f(x)

x = torch.arange(17)
torch._dynamo.mark_dynamic(x, 0)
out = f(x)
```

cc @chauhang @penguinwu @ezyang @bobrenjc93 @zou3519

**Comment by ProExpertProg** [comment]
Thanks! This would benefit vLLM `torch.compile` usage, currently we have to turn padding off during compilation.

**Comment by laithsakka** [comment]
so  max(x.shape[0], 16) should get converted sym_max = torch.sym_max(s0, 16); in the graph. 

I agree there seem to be over guarding that can be avoided, we probably should debug this
I tried this program:
```
@torch.compile(fullgraph=True)
def f(x):
    padded_size = max(x.shape[0], 16)
    return x,padded_size

x = torch.arange(15)
torch._dynamo.mark_dynamic(x, 0)
out = f(x)
x = torch.arange(20)

```
with dynamic we get a recompilation
  
https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/lsakka/c6e673df-035c-4a6e-8184-40dcf1cdb5cb/custom/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000

BUT if we use mark_unbacked it actually works single graph meaning that in theory guards are not needed. and its actually clear from the graphs that there is no reason for the guards.
they should be added only if some one try to materialize the max symbolic expression

https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/lsakka/e24b670e-4ef1-418e-b790-fba5c01c4c62/custom/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000

**Comment by laithsakka** [comment]
turn out we explicitly specialize sym_max and sym_min not sure if this was intended or just old cold. 
here a test for the failing tests then we can decide what next steps 
https://github.com/pytorch/pytorch/pull/149656

**Comment by laithsakka** [comment]
tried it on the example above and it does not generate recompilations

**Comment by laithsakka** [comment]
ok there are some test failures we need to look at those to see if we hard dependency on such specialization for those or not.
, I will remove my self assignment to this for now since i do not have bandwidth for this short term will look again when i have time maybe in two weeks if now one else picked it up.

**Comment by laithsakka** [comment]
related to this 
    k = min(m, n)
with unbacked inputs will throw data dependent error even when its not needed to be materialized

**Comment by laithsakka** [comment]
I see the logic of this guarding was added here by
https://github.com/pytorch/pytorch/pull/94400/files
@tugsbayasgalan wonder if you were able to knock of all faluires.
